lib/api/objectPutPart.js

```diff
@@ -87,80 +56,146 @@ export default function objectPutPart(authInfo, request, streamingV4Params,
-                    const serverSideEncryption =
-                              bucket.getServerSideEncryption();
-                    if (serverSideEncryption) {
-                        return kms.createCipherBundle(
-                            serverSideEncryption, log, (err, cipherBundle) => {
-                                services.metadataValidateMultipart(
-                                    metadataValMPUparams,
-                                    doStore(objectKeyContext, cipherBundle,
-                                            request, log, paddedPartNumber,
-                                            size, uploadId,
-                                            streamingV4Params, cb));
-                            });
+                    const initiatorID = res.initiator.ID;
+                    const requesterID = authInfo.isRequesterAnIAMUser() ?
+                        authInfo.getArn() : authInfo.getCanonicalID();
+                    if (initiatorID !== requesterID) {
+                        return next(errors.AccessDenied);
                     }
-                    return services.metadataValidateMultipart(
-                        metadataValMPUparams,
-                        doStore(objectKeyContext, null, request, log,
-                                paddedPartNumber, size, uploadId,
-                                streamingV4Params, cb));
+                    return next();
                 });
-        });
+        },
+        // Store in data backend.
+        next => {
+            const objectKeyContext = {
+                bucketName,
+                owner: canonicalID,
+                namespace: request.namespace,
+            };
+            return services.dataStore(null, objectKeyContext, cipherBundle,
+                request, size, streamingV4Params, log,
+                (err, objectMetadata, dataGetInfo, hexDigest) =>
+                    next(err, dataGetInfo, hexDigest));
+        },
+        // Store data locations in metadata.
+        (dataGetInfo, hexDigest, next) => {
+            // Use an array to be consistent with objectPutCopyPart where there
+            // could be multiple locations.
+            const dataGetInfoArr = [dataGetInfo];
+            if (cipherBundle) {
+                const { algorithm, masterKeyId, cryptoScheme,
+                    cipheredDataKey } = cipherBundle;
+                dataGetInfoArr[0].sseAlgorithm = algorithm;
+                dataGetInfoArr[0].sseMasterKeyId = masterKeyId;
+                dataGetInfoArr[0].sseCryptoScheme = cryptoScheme;
+                dataGetInfoArr[0].sseCipheredDataKey = cipheredDataKey;
+            }
+            const mdParams = {
+                partNumber: _getPaddedPartNumber(partNumber),
+                contentMD5: hexDigest,
+                size,
+                uploadId,
+                splitter,
+            };
+            return services.metadataStorePart(mpuBucketName, dataGetInfoArr,
```

Comment:
This can be extracted into this api as well. It doesn't carry any extra setup, so it should be ok.